### ISSUE TITLE
[Deliverable] Deploy RLN smart contract to a L2 mainnet

### ISSUE THREAD

**Issue by chair28980** [waku-org/pm#258]
[Roadmap definition](https://roadmap.logos.co/waku/milestones/2024-rln-mainnet#deliverable-deploy-rln-smart-contract-to-a-l2-mainnet)

## Output

---

## Background

The RLN contract is a key component of Waku architecture, and a new version is currently in development. This upgrade aims to enhance membership management capabilities, as outlined in the [specification](https://github.com/waku-org/specs/blob/master/standards/core/rln-contract.md).

The current RLN contract is deployed only on the Sepolia testnet, but the upgraded version is planned for deployment on a mainnet, specifically Linea, a ZK-rollup.

This document formalizes the RLN contract upgrade process. While motivated by the current upgrade, the outlined steps are general enough to be reused for future upgrades.

The contract rollout procedure involves at least two deployments: first to a testnet, and later to a mainnet. These steps may be repeated as needed based on testing outcomes. The upgrade procedure outlines the minimal necessary steps but can be expanded if required.

### 1. Setup a Multisig

The contract should not be controlled by a single party. Ownership must eventually be renounced, and, in the meantime, distributed through a multisig contract. Key steps include:

- Choose a multisig implementation, like [Safe](https://safe.global/wallet), and evaluate alternatives and third-party risks.
- Determine a quorum configuration (e.g., 2-of-3) and assess if it meets our needs.
- Assign signers—team members for testnet and possibly external parties for mainnet.
- Deploy the multisig wallet and test signer control of keys.
- Establish a key rotation procedure in case of lost or stolen keys.

### 2. Deploy the Contract

- Acquire test tokens from [Linea Sepolia faucets](https://docs.linea.build/users/move-funds/fund#linea-sepolia) (or bridge ETH to Linea for mainnet).
- Deploy the contract to the chosen network (testnet or mainnet), and back up all relevant deployment information (bytecode, compiler version, deployment transaction hash).
- For mainnet, verify the contract on Etherscan or similar services.

### 3. Upgrade nwaku

- Update nwaku to point to the new RLN contract.
- Upgrade nwaku-compose to use this new version of nwaku.

### 4. Upgrade RLN Membership Registration Tools

Modify RLN membership registration tools, such as the `register-rln.sh` script (see [nwaku-compose](https://github.com/waku-org/nwaku-compose)), to reflect the new contract and allow user-specified rate limits. Optionally, extend functionality to support additional contract actions like "extend membership" and "withdraw deposit."

### 5. Test RLN Functionality

Test the upgraded contract using the registration tools, addressing any issues and optimizing for gas efficiency.

### 6. Transfer Ownership to the Multisig

Transfer contract ownership to the multisig, balancing the timing between sufficient testing and minimizing risks.

### 7. Upgrade The Waku Network (TWN)

Encourage TWN node operators to upgrade their nodes. Steps include:

- Upgrade our own nwaku nodes and test them in a P2P setting (cf. waku-simulator).
- Write and distribute upgrade instructions to node operators (blog post, Discord, etc).
- Announce the upgrade with a clear timeline (including a time point when the old memberships must be considered invalid).
- Collect and address feedback from node operators.

### 8. Review and Audit the Contract

Conduct both internal reviews and external audits throughout the process, especially for the final mainnet version.

### 9. Update the RLN Contract Specification

As the upgrade reaches mainnet, update the [specification](https://github.com/waku-org/specs/blob/master/standards/core/rln-contract.md) to reflect the upgraded mainnet contract. Advance the specification to "draft" status.

## Future Work

The following steps may be considered for medium-to-long term after the upgrade is complete.

### 1. Contract Monitoring

Implement real-time monitoring and alerts for unusual activity.

### 2. Bug Bounty Program

Launch a bug bounty program to identify and address vulnerabilities.

### 3. Testnet-TWN Parallel to Mainnet-TWN

The current upgrade will move TWN from a testnet contract on Sepolia (L1) to a mainnet contract on Linea L2.
Consider maintaining a testnet-based Waku network to test future upgrades in a realistic setting.

### 4. UI-based RLN Membership Tools

Develop a user-friendly interface to manage RLN memberships, displaying key membership data and enabling actions like extending or withdrawing memberships.

**Comment by s-tikhomirov** [comment]
Added an outline for the RLN contract rollout procedure on Linea (testnet, then mainnet).

**Comment by jm-clius** [comment]
Thanks for the breakdown! My interpretation of the next tasks and possible owners (perhaps we can create subissues for each, if you agree)?

1. Deploy contract to Linea testnet (already being tracked here: https://github.com/waku-org/waku-rlnv2-contract)
Owner: @richard-ramos

2. Update nwaku-compose & registration tools to work with Linea testnet contract (covers 3,4,5,7 for Linea testnet above)
Owner: Not sure, @richard-ramos perhaps this can also be something owned by you and nwaku team post contract deployment?

3. Setup multisig and transfer ownership
Owner: Perhaps @s-tikhomirov and Research team, with input from rest of IFT

Future tasks will then involve the audit, moving to mainnet, upgrading the spec etc. I think we can create issues for these when we get to this point.

**Comment by s-tikhomirov** [comment]
LGTM!

> Setup multisig and transfer ownership
> Owner: Perhaps @s-tikhomirov and Research team, with input from rest of IFT

Is this task testnet-only initially, in your view? I'm not sure how much more stringent we need to be with the multisig on testnet as compared to mainnet...

**Comment by richard-ramos** [comment]
@jm-clius:  
> 2. Update nwaku-compose & registration tools to work with Linea testnet contract (covers 3,4,5,7 for Linea testnet above)
> perhaps this can also be something owned by you and nwaku team post contract deployment

cc-ing @Ivansete-status for his thoughts but I think this is the plan!

**Comment by richard-ramos** [comment]
For testnet I think it makes sense to have a multisig with at least 3 participants, but only requiring a single signature to authorize transactions (to not delay changes on testnet if required). 

This is to minimize the likelyhood of having to deploy a new contract on the testnet, update nwaku-compose and ask nwaku users to register in this new contract.  Such a scenario could arise due to various reasons, including the potential loss of the keys associated with the address that owns the smart contract; the owner of the contract leaving Waku without designating a new owner for the contract, or, unfortunately, unforeseen circumstances such as a serious accident that would stop the owner from setting up a new owner for the contract.

**Comment by Ivansete-status** [comment]
> [@jm-clius](https://github.com/jm-clius):
> 
> > 2. Update nwaku-compose & registration tools to work with Linea testnet contract (covers 3,4,5,7 for Linea testnet above)
> >    perhaps this can also be something owned by you and nwaku team post contract deployment
> 
> cc-ing [@Ivansete-status](https://github.com/Ivansete-status) for his thoughts but I think this is the plan!

Thanks for the heads-up @richard-ramos ! In my opinion, the RLN-related work has the lowest priority. I'd vote to start working on that in H2-2025, or at least after we've completed the integration of nwaku in status.

wdyt @jm-clius , @fryorcraken ?

**Comment by Ivansete-status** [comment]
Okay, thanks for clarifying @jm-clius ! If it is a small task we can interleave it with the status-go integration.
When would you like this task to be completed?

**Comment by fryorcraken** [comment]
> 1. Setup a Multisig

It would be good to also summarize the rights of the _owner_ of the contract for the _implementation_ somewhere (README of contract repo?) so that it's clearly stated when discussing with rest of the org.

**Comment by s-tikhomirov** [comment]
> summarize the rights of the owner of the contract for the implementation somewhere (README of contract repo?)

Submitted a PR:

https://github.com/waku-org/waku-rlnv2-contract/pull/23

**Comment by jm-clius** [comment]
Note that steps 3 and 4 requires the client changes scoped as part of https://github.com/waku-org/pm/issues/228

**Comment by chair28980** [comment]
Duplicate issue https://github.com/waku-org/pm/issues/230

**Comment by fryorcraken** [comment]
@jm-clius Do we agree that we should reduce the scope of this and limit it to a Linea-based L2 testnet aka Linea Sepolia?

**Comment by fryorcraken** [comment]
After review, we will not aim for mainnet just yet for a few reasons:
1. we are polishing the delivery of onchain tree mgmt. This needs to be deployed and dogfed before we can assert we have a stable contract for audit/mainnet
2. we need to iterate more on UX and security, onchain tree mgmt (1) improves UX a lot, but we have already spotted other potential improvements (slashing, multiple roots on chain). Software needs to be used more before we can assess
3. Ideally, we would go directly to Status L2 mainnet. Now planned for 2027.

See https://github.com/waku-org/pm/issues/286 for deployment on Linea Testnet.

cc @jm-clius @chair28980
